src/Elastic.Apm.AspNetCore/ApmMiddleware.cs

```diff
@@ -42,21 +42,38 @@ public ApmMiddleware(RequestDelegate next, Tracer tracer, IApmAgent agent)
 
 		public async Task InvokeAsync(HttpContext context)
 		{
-			var transaction = StartTransaction(context);
+			var transaction = await StartTransactionAsync(context);
 
 			try
 			{
+				// We need to reset the current transaction (which is async local), since the StartTransactionAsync call above ended with all its
+				// child spans at this point, therefore need to set the current transaction here again.
+				_tracer.CurrentExecutionSegmentsContainer.CurrentTransaction = transaction;
 				await _next(context);
 			}
-			catch (Exception e) when (transaction != null && ExceptionFilter.Capture(e, transaction, context, _logger))
-			{ }
+			catch (Exception e) when (transaction != null)
+			{
+				transaction.CaptureException(e);
+				// It'd be nice to have this in an exception filter, but that would force us capturing the request body synchronously.
+				// Therefore we rather unwind the stack in the catch block and call the async method.
+				if (context != null && transaction.IsCaptureRequestBodyEnabled(true))
+					await transaction.CollectRequestBody(true, context.Request, _logger);
+
+				throw;
+			}
 			finally
 			{
+				// In case an error handler middleware is registered, the catch block above won't be executed, because the
+				// error handler handles all the exceptions - in this case, based on the response code and the config, we may capture the body here
+				if (transaction != null && context?.Response.StatusCode >= 300 && transaction.Context?.Request?.Body is string body

```

Comment:
Is "not generating error when error handler middleware is registered" an issue in ASP.NET Core 2.0 + "agent before this PR's changes" as well? Is this change related to the purpose of this PR?

Reply:
Sorry, I don't get the first part of the question. 

> Is this change related to the purpose of this PR?

Good point, I appreciate we follow questioning these thing! 
Yes, it is very much related. I had a rant about it [here](https://github.com/elastic/apm-agent-dotnet/pull/608#discussion_r347429628). Since we read the request body asynchronously we can't do it in the `void AspNetCoreDiagnosticListener.OnNext(...)` method anymore - so the code that reads the body in case an error happened and is handled by an error handling middleware is moved here from the `AspNetCoreDiagnosticListener.OnNext` method.

Reply:
I see. I have two points of concern:
1. This effectively changes the definition of an error. For example, do we want to treat HTTP status code 304 (Not Modified) as an error?
2. I assume `AspNetCoreDiagnosticListener.OnNext` is called before we get to this point of `ApmMiddleware.InvokeAsync` which means error event's context will not have request body. If we are going to align with other agents and send request body in error event's context even for non-sampled transaction wouldn't it be a problem?

Is there a reason you didn't take approach mentioned in [the comment](https://github.com/elastic/apm-agent-dotnet/pull/608#discussion_r347429628) 

> Another option would be to mark the transaction here and not do the reading and have a check in 

Generating error event detected in `AspNetCoreDiagnosticListener` can be deferred to this point of the flow thus solving both points above.

Reply:
I think handling `context?.Response.StatusCode >= 300` as error was really not a good idea, I changed it to `context?.Response.StatusCode >= 400`. 

To your points:
1. After doing the change mentioned above, not anymore. What typically an error handling middleware does is that it handles the exception, sets the return code to some http 4xxx or 5xx and returns an HTML page.
2. Right, that makes it more challenging, I think what you say in your last sentence is a good approach. Nevertheless, I would not worry about it now, this is not something that the PR changes, this behaviour remains the same.

> Is there a reason you didn't take approach mentioned in the comment

I did not find a good way to do it - like, adding a flag to transaction and things like that does not sound very nice to me, and this approach looks cleaner.

Long term, when we align on the error context capturing behaviour we can revisit this, but for now I'd leave that out from this PR.

Reply:
> I did not find a good way to do it - like, adding a flag to transaction and things like that does not sound very nice to me, and this approach looks cleaner.

I'm not sure what are the cons of adding a flag to transaction but please note that this change does break existing functionality - before this change error event [captured in `AspNetCoreDiagnosticListener.OnNext`](https://github.com/elastic/apm-agent-dotnet/pull/608/files#diff-178ddb4c5da109db16c57afee4635d9aR29) had request body (assuming configuration allowed it and transaction is sampled) but after this change it sometimes will and sometimes it won't. The tests pass because they check error event context after transaction is completed (and because we have this issue of not taking a snapshot of transaction context for error event) but in real life use it will be a data race between transaction thread and payload sender's thread - if error event is serialized to JSON before transaction completes serialized error event won't have request body.
I leave it to you to decide if the severity of this issue warrants additional work.

Reply:
Oh, I see what you mean, and yes that change is true indeed. But that code was kind of already ignoring this whole thing I feel (mainly because I think the error's context was not really thought through I feel).

We did basically this:
- Create an error and then set its context pointing to the transactions context
- Then, in a totally different call, set the body for the transaction's context, not thinking anything about the error's. 

I'd just do another iteration on when and how we capture errors context - and remove the sharing of that instance, I think that's worth a whole new PR.